Phase2 Outer Tracker Clusterizer integration in CMSSW 8

thomaslenzi opened the pull request:
This PR migrates the Phase2 Outer Tracker Clusterizer from CMSSW 6_2_SLHC to CMSSW 8. It adds the new data format for the clusters as well as the clustering algorithm itself.

Commit:
First update of the Clusterizer to CMSSW 8'

Commit:
Further migration of clusterizer

cmsbuild commented:
A new Pull Request was created by  @thomaslenzi (Thomas Lenzi) for CMSSW_8_0_X.

It involves the following packages:

DataFormats/Phase2TrackerCluster
RecoLocalTracker/SiPhase2Clusterizer


The following packages do not have a category, yet:

RecoLocalTracker/SiPhase2Clusterizer

@cmsbuild, @cvuosalo, @davidlange6, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @GiacomoSguazzoni, @rovere, @VinInn, @gpetruc, @cerati, @threus this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

thomaslenzi commented:
I would also like to note that I have informed Daniel Elvira about the new package RecoLocalTracker/SiPhase2Clusterizer

slava77 commented on DataFormats/Phase2TrackerCluster/src/classes_def.xml:
```diff
@@ -0,0 +1,12 @@
+<lcgdict>
+    <class name="Phase2TrackerCluster1D" ClassVersion="10">
```
we recommend to start from version 2 for new classes

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/plugins/BuildFile.xml:
```diff
@@ -0,0 +1,4 @@
+<use name="RecoLocalTracker/SiPhase2Clusterizer"/>
+<library   file="Phase2TrackerClusterizer.cc ../src/Phase2TrackerClusterizerAlgorithm.cc ../src/Phase2TrackerClusterizerArray.cc" name="Phase2TrackerClusterizerPlugin">
```
either move the code from src/ to plugins/ or let the parts live  in separate libraries.

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/plugins/Phase2TrackerClusterizer.cc:
```diff
@@ -0,0 +1,89 @@
+#include <vector>
+#include <iostream>
+
+namespace cms {
```
namespace "cms" is not particularly necessary in this case; ~ OK to keep.

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/plugins/Phase2TrackerClusterizer.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace cms {
+
+    class Phase2TrackerClusterizer : public edm::EDProducer {
```
new modules should be thread-aware.
Use edm::stream or edm::global  (it looks like edm::global::EDProducer will work)

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/plugins/Phase2TrackerClusterizer.cc:
```diff
@@ -0,0 +1,89 @@
+            // Objects that will be produced
+            produces< Phase2TrackerCluster1DCollectionNew >(); 
+            // 
+            clusterizer_  = new Phase2TrackerClusterizerAlgorithm(maxClusterSize_, maxNumberClusters_);
```
use std::unique_ptr instead (no need to delete in this case).
token_ and clusterizer_ could be initialized in the initializer list (between ":" and "{" above.

delaere commented on RecoLocalTracker/SiPhase2Clusterizer/src/Phase2TrackerClusterizerAlgorithm.cc:
```diff
@@ -0,0 +1,96 @@
+
+void Phase2TrackerClusterizerAlgorithm::clusterizeDetUnit(const edm::DetSet< Phase2TrackerDigi >& digis, edmNew::DetSetVector< Phase2TrackerCluster1D >::FastFiller& clusters) {
+
+    // Fill the 2D matrix with the ADC values
```
Please change the comment to avoid misunderstandings... there is no more ADC values.

slava77 commented:
@thomaslenzi 
it will take me a bit longer to send all comments to the code, I'll post a message when I'm done

slava77 commented:
@cmsbuild please test
to collect other possible issues with the code

cmsbuild commented:
The tests are being triggered in jenkins.

Commit:
Applied comments from PR #13087 for the clusterizer integration in CMSSW 8

cmsbuild commented:
Pull request #13087 was updated. @cmsbuild, @cvuosalo, @davidlange6, @slava77 can you please check and sign again.

thomaslenzi commented:
I pushed changes to my branch and hope this will solve the problems highlighted here-above.

cmsbuild commented:
+1
Tested at: 81972931a251b93b54b45766684257e8cccf80af
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10779/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10779/summary.html

Commit:
Added RecHits to the PR for CMSSW_8_0_X

cmsbuild commented:
Pull request #13087 was updated. @cmsbuild, @cvuosalo, @davidlange6, @slava77 can you please check and sign again.

thomaslenzi commented:
I added the changes of PR #13099 to this PR as suggested.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef DATAFORMATS_PHASE2TRACKERCLUSTER_PHASE2TRACKERCLUSTER1D_H 
+#define DATAFORMATS_PHASE2TRACKERCLUSTER_PHASE2TRACKERCLUSTER1D_H 
+
+#include <stdint.h>
```
@Dr15Jones 
should we be moving to cstdint (C++ header) instead for the fixed width types?

Dr15Jones replied:
@pcanal We should move to the fixed width headers. However, I do not know how ROOT interacts with them in classes that are meant to be stored.

cmsbuild commented:
+1
Tested at: b408a53ccfb6d883f971a9a5ba135987665e1131
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10815/summary.html

slava77 commented on DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h:
```diff
@@ -0,0 +1,43 @@
+
+private:
+
+    Phase2TrackerDigi firstDigi_;
```
Logically, it's not the best way to indicate the first channel.
It may be better to define a Phase2TrackerChannelId instead, in case Digi develops into something more (e.g. some channel quality bits get added).
... probably still OK for now.

delaere replied:
It is quite unlikely for the Phase2TrackerDigi to evolve into something more complex. The digi class was developed together with the unpacker and the raw dataformat is more or less frozen. The main reason to use Phase2TrackerDigi was to ensure consistency in the underlying dataformat, but that's not a strong argument.

slava77 commented on DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h:
```diff
@@ -0,0 +1,43 @@
+    uint16_t size() const { return (data_ & 0x7fff); }
+    uint16_t threshold() const { return ((data_ >> 15) & 0x1); }
+    float center() const { return firstStrip() + (data_ & 0x7fff) / 2.; }
+    std::pair< float, float > barycenter() const { return std::make_pair(column(), center()); }
```
I suppose .first for the z (localy) direction and .second for x are not confusing.
Also, is it meaningful to define the second dimension for the barycenter of a 1D type .. and also to convert it to a float already here?

delaere replied:
These are 1D clusters built in a 2D plane, so the second coordinate actually makes sense. We do 1D because that's what the electronics does.

slava77 commented on DataFormats/Phase2TrackerRecHit/interface/Phase2TrackerRecHit1D.h:
```diff
@@ -0,0 +1,33 @@
+
+typedef edm::Ref< edmNew::DetSetVector< Phase2TrackerCluster1D >, Phase2TrackerCluster1D > Phase2ClusterReference;
+
+class Phase2TrackerRecHit1D {
```
no connection to TrackingRecHit?
Why not? was there an issue with persistence or something else?

lowette replied:
We just added something minimal. We are far from connecting to the tracking.

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/interface/Phase2StripCPEDummy.h:
```diff
@@ -0,0 +1,20 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+
+class Phase2StripCPEDummy : public ClusterParameterEstimator<Phase2TrackerCluster1D> {
```
maybe Phase2StripTrivialCPE
instead of "Dummy"?
Trivial appears to be more commonly used for something producing results with very simple logic;
a Dummy is more for something that does nothing at all.

lowette replied:
Sounds good.

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2StripCPEESProducer.cc:
```diff
@@ -0,0 +1,34 @@
+}
+
+
+boost::shared_ptr<ClusterParameterEstimator<Phase2TrackerCluster1D> > Phase2StripCPEESProducer::produce(const TkStripCPERecord & iRecord) {
```
time to move to std::shared_ptr

Dr15Jones replied:
I don't know if we yet support ```std::shared_ptr``` for ```EventSetup``` producers.

Dr15Jones replied:
I checked and we haven't supported it yet.

slava77 replied:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/FWCore/Skeletons/scripts/mkTemplates/ESProducer/ESProducer.cc
is the template (updated in 2014) incorrect?

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2StripCPEESProducer.h:
```diff
@@ -0,0 +1,36 @@
+#include "RecoLocalTracker/Records/interface/TkStripCPERecord.h"
+#include "RecoLocalTracker/ClusterParameterEstimator/interface/ClusterParameterEstimator.h"
+#include "DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h"
+#include <boost/shared_ptr.hpp>
```
move to std::shared_ptr and use ``` <memory>```

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2TrackerRecHits.h:
```diff
@@ -0,0 +1,34 @@
+#include "RecoLocalTracker/Phase2TrackerRecHits/interface/Phase2StripCPEDummy.h"
+
+
+  class Phase2TrackerRecHits : public edm::global::EDProducer<> {
```
it is actually not worth to make a header file for this small piece, it can be fully embedded in Phase2TrackerRecHits.cc
.. similar story for a few other plugins/\*.h files, which are meant to be included by only one .cc

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10815/summary.html

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/src/Phase2StripCPEDummy.cc:
```diff
@@ -0,0 +1,17 @@
+
+
+Phase2StripCPEDummy::LocalValues Phase2StripCPEDummy::localParameters(const Phase2TrackerCluster1D & cluster, const GeomDetUnit & det) const {
+  float strippitch  = 0.0090; // hardcoded dummy, a la 2S
```
Will these be used really only for 2S modules?
PS-s part is 0.01x2.5 .. and PS-p is probably not covered by this CPE (or it can be, but then the length is 0.15)
Please clarify.

lowette replied:
It is a placeholder while we bring in the geometry dependence in a next step. It really doesn't matter what the exact value is.

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/test/RecHitsValidation.cc:
```diff
@@ -0,0 +1,276 @@
+
+        explicit Phase2TrackerRecHitsValidation(const edm::ParameterSet&);
+        ~Phase2TrackerRecHitsValidation();
+        void beginJob();
```
``` override ``` 
would be useful for all supposedly inherited methods

lowette replied:
Looks like this one is worth learning about. I'm getting old...

slava77 replied:
Methods like beginJob is something you'd expect to be called by the framework when needed. However, if the base class method changes name or signature, your function defined here may never be called.
Add  ``` override ``` to tell the compiler this has to be also in the base class.

the ED\* workers interface changed a few times in the past years. So, better learn early that the method you implemented actually does something in run time.

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/test/RecHitsValidationTest_cfg.py:
```diff
@@ -0,0 +1,68 @@
+
+process.schedule = cms.Schedule(process.rechits_step, process.validation_step)
+
+from SLHCUpgradeSimulations.Configuration.phase2TkCustomsBE5DPixel10Ddev import customise_condOverRides
```
does the phase2TkCustomsBE5DPixel10Ddev file exist?
It would be nice to have operational tests.

lowette replied:
Sure it would be nice, but we don't even have the digis yet. We're not at that stage yet, and this can be cleaned up when we reach there.

slava77 replied:
well, what's the point to put config files that can not be run?

slava77 commented on RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2TrackerRecHits.cc:
```diff
@@ -0,0 +1,79 @@
+  std::auto_ptr< Phase2TrackerRecHit1DCollectionNew > outputRecHits(new Phase2TrackerRecHit1DCollectionNew());
+
+  // Loop over clusters
+  for (edmNew::DetSetVector< Phase2TrackerCluster1D >::const_iterator DSViter = clusters->begin(); DSViter != clusters->end(); ++DSViter) {
```
``` for (auto clu : clusters) ```
may be less verbose and easier to read

lowette replied:
I as good as stopped coding before C++11 came in. Not sure it is worth learning-and-forgetting-again.

slava77 replied:
I think this is one of those things worth learning.
Instead of what you have above you can have
``` for (auto clu : *clusters){ 
     DetId detId(clu.detId()); 
...
``` 
with a smaller number of symbols to type or make typos in.

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/interface/Phase2TrackerClusterizerAlgorithm.h:
```diff
@@ -0,0 +1,34 @@
+
+        Phase2TrackerClusterizerAlgorithm(unsigned int, unsigned int);
+        void setup(const PixelGeomDetUnit*);
+        void clusterizeDetUnit(const edm::DetSet< Phase2TrackerDigi >&, edmNew::DetSetVector< Phase2TrackerCluster1D >::FastFiller&);
```
here and in a few more places which already have Phase2TrackerCluster1D included
may want to use ``` Phase2TrackerCluster1DCollectionNew ``` instead of 
``` edmNew::DetSetVector< Phase2TrackerCluster1D >```

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/test/BuildFile.xml:
```diff
@@ -0,0 +1,28 @@
+<use   name="DataFormats/ParticleFlowReco"/>
+<library   file="ClustersValidation.cc" name="Phase2TrackerClusterizerValidation">
+  <flags   EDM_PLUGIN="1"/>
+  <flags CXXFLAGS="-g3"/>
```
this is a /test build, still, the debug flags are not supposed to be set

slava77 commented on RecoLocalTracker/SiPhase2Clusterizer/test/ClustersValidation.cc:
```diff
@@ -0,0 +1,497 @@
+
+        explicit Phase2TrackerClusterizerValidation(const edm::ParameterSet&);
+        ~Phase2TrackerClusterizerValidation();
+        void beginJob();
```
add ``` override ``` just to be sure the methods are actually doing anything

slava77 commented:
@thomaslenzi 
I'm done with the review at this point.
It would be good to get some feedback or follow up to most of the comments above.
Please suggest a sequence of actions to test the code run time.

Also, I'd like to ask for a presentation in an upcoming RECO meeting (our next slot is Feb 3, but it is still to be confirmed)

slava77 commented:
Based on the feed back so far, it sounds like this code can not be tested beyond compilation because there are no inputs to try. 
Is this correct?
(BTW, HGCAL has running tests in 80x).

delaere commented:
The code was tested in SLHCDEV.  Tests in 80X require a PR for the digitizer expected this week. This will also be included in the 2023LReco workflow as part of the tracker phase2 eras definition soon after.  I will sketch the complete plan tomorrow at the simulation meeting.

slava77 commented:
On 1/28/16 4:16 PM, Christophe Delaere wrote:
> The code was tested in SLHCDEV. 

Ah, good to know.
This is much less abstract then.


Please clarify if you can update the code based on comments in this PR.

Thank you



> Tests in 80X require a PR for the
> digitizer expected this week. This will also be included in the
> 2023LReco workflow as part of the tracker phase2 eras definition soon
> after. I will sketch the complete plan tomorrow at the simulation meeting.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/13087#issuecomment-176493065>.
>

Commit:
Moved from stdint.h to cstdint
Changed "Dummy" for "Trivial"
Tried to move to "std::shared_ptr" but generated errors so used "boost:shared_ptr"
Combined .h and .cc files of some of the classes
Added "override" statement to EDAnalyzers
Removed the "phase2TkCustomsBE5DPixel10Ddev" call in the python files
Use "auto" to loop over elements
Changed "edmNew::DetSetVector< Phase2TrackerCluster1D >" for Phase2TrackerCluster1DCollectionNew
Removed <flags CXXFLAGS="-g3"/>

cmsbuild commented:
Pull request #13087 was updated. @cmsbuild, @cvuosalo, @davidlange6, @slava77 can you please check and sign again.

thomaslenzi commented:
@slava77 I applied the modifications as you suggested and tried to do so in all the files (eg use auto also for the clusters, ...). Here is the list of what I changed:

Moved from stdint.h to cstdint
Changed "Dummy" for "Trivial"
Tried to move to "std::shared_ptr" but generated errors so used "boost:shared_ptr"
Combined .h and .cc files of some of the classes
Added "override" statement to EDAnalyzers
Removed the "phase2TkCustomsBE5DPixel10Ddev" call in the python files
Use "auto" to loop over elements
Changed "edmNew::DetSetVector< Phase2TrackerCluster1D >" for Phase2TrackerCluster1DCollectionNew
Removed <flags CXXFLAGS="-g3"/>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8628a59f32ec06d06efdee5a47b904a28ffb38df
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10845/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13087/10845/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/13087	 8628a59
- code changes look somewhat OK, considering the preliminary stages of integration; only new code is added
- jenkins tests pass: compilation is OK and the static analyzer shows no issues (although I'm not fully sure it checks the new packages); comparisons with baseline show no differences (more as a sanity check).

smuzaffar commented:
cms-bot recheck packages

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

VinInn commented:
@davidlange6, what is the status of this?
moved to 810 or will be integrated in some sort of 801?

davidlange6 commented:
sorry, is in need of me to look - but would prefer its for 81x.

> On Mar 1, 2016, at 4:38 PM, Vincenzo Innocente <notifications@github.com> wrote:
> 
> @davidlange6, what is the status of this?
> moved to 810 or will be integrated in some sort of 801?
> 
> —
> Reply to this email directly or view it on GitHub.
>

boudoul commented:
It is in 800pre6 , isnt' it ?
https://github.com/cms-sw/cmssw/releases/CMSSW_8_0_0_pre6

VinInn commented:
oh yes, sorry did not noticed that was merged on January 31!

delaere commented:
Yes indeed. This one was merged. The question is valid for PR #13185 vs #13368 (the digitizer). In principle the digitizer is needed by the clusterizer, but I have no problem if we skip 80X, as we have presently no phase2tk workflow in that release cycle.

davidlange6 commented:
indeed, I missed clusterizer->digitizer… I need one last look at the digitizer PR and it goes for 81x (and phase2 should forget about 80x for now)

> On Mar 1, 2016, at 4:58 PM, Christophe Delaere <notifications@github.com> wrote:
> 
> Yes indeed. This one was merged. The question is valid for PR #13185 vs #13368 (the digitizer). In principle the digitizer is needed by the clusterizer, but I have no problem if we skip 80X, as we have presently no phase2tk workflow in that release cycle.
> 
> —
> Reply to this email directly or view it on GitHub.
>

VinInn commented:
yes. And we need to proceed forward. We need the digitizer and new workflows so that tracking can plugin and proceed as well

boudoul commented:
the workflow with phase2 (flat and tilted) up to gen sim just pull requested in 81X - Including the local reco steps will follow (factorizing the 2 steps to ease the integration)
